internal/v2/application/command.go

```diff
@@ -0,0 +1,735 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package application
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
+	sdkCommon "github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/container"
+	"github.com/edgexfoundry/device-sdk-go/internal/transformer"
+	dsModels "github.com/edgexfoundry/device-sdk-go/pkg/models"
+	bootstrapContainer "github.com/edgexfoundry/go-mod-bootstrap/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/di"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/coredata"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	edgexErr "github.com/edgexfoundry/go-mod-core-contracts/errors"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/dtos"
+)
+
+type CommandProcessor struct {
+	device         *contract.Device
+	deviceResource *contract.DeviceResource
+	id             string
+	cmd            string
+	params         string
+	dic            *di.Container
+}
+
+func NewCommandProcessor(device *contract.Device, dr *contract.DeviceResource, id string, cmd string, params string, dic *di.Container) *CommandProcessor {
+	return &CommandProcessor{
+		device:         device,
+		deviceResource: dr,
+		id:             id,
+		cmd:            cmd,
+		params:         params,
+		dic:            dic,
+	}
+}
+
+func CommandHandler(isRead bool, sendEvent bool, id string, vars map[string]string, body string, dic *di.Container) (event *dtos.Event, err edgexErr.EdgeX) {
+	var device contract.Device
+	deviceKey := vars[sdkCommon.IdVar]
+
+	defer func() {
+		// update LastConnected when the device was successfully interacted with
+		if err == nil {
+			go sdkCommon.UpdateLastConnected(
+				device.Name,
+				container.ConfigurationFrom(dic.Get),
+				bootstrapContainer.LoggingClientFrom(dic.Get),
+				container.MetadataDeviceClientFrom(dic.Get))
+		}
+		// push event to coredata if specified
+		if sendEvent {
```

Comment:
I think this needs to only be called if `err == nil`, so move up into previous if block.

Or better yet,  just put the following at the top of this defered func to avoid indentation. 

```
if err != nil {
   return
}
```